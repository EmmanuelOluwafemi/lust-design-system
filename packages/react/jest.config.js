module.exports = {
    "testEnvironment": "jsdom",
    roots: ['<rootDir>/src'],
    textRegex: '(/.*\\.test)\\.(ts|tsx)$',
    setupFilesAfterEvn: ['<rootDir>/jest.setup.ts']
}