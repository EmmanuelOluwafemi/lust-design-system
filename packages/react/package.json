{
  "name": "@lust/react",
  "version": "1.0.0",
  "main": "lib/index.js",
  "license": "MIT",
  "devDependencies": {
    "@babel/core": "^7.18.5",
    "@babel/preset-env": "^7.18.2",
    "@babel/preset-react": "^7.17.12",
    "@babel/preset-typescript": "^7.17.12",
    "@storybook/addon-a11y": "^6.5.9",
    "@storybook/addon-actions": "^6.5.9",
    "@storybook/addon-essentials": "^6.5.9",
    "@storybook/addon-interactions": "^6.5.9",
    "@storybook/addon-knobs": "^6.4.0",
    "@storybook/addon-links": "^6.5.9",
    "@storybook/addon-storysource": "^6.5.9",
    "@storybook/builder-webpack4": "^6.5.9",
    "@storybook/manager-webpack4": "^6.5.9",
    "@storybook/preset-typescript": "^3.0.0",
    "@storybook/react": "^6.5.9",
    "@storybook/testing-library": "^0.0.13",
    "@testing-library/jest-dom": "^5.16.4",
    "@testing-library/react": "^13.3.0",
    "@types/jest": "^28.1.3",
    "@types/react": "^18.0.12",
    "babel-loader": "^8.2.5",
    "chromatic": "^6.6.3",
    "jest": "^28.1.1",
    "jest-environment-jsdom": "^28.1.1",
    "react": "^18.1.0",
    "rollup": "^2.75.6",
    "rollup-plugin-typescript2": "^0.32.1",
    "typescript": "^4.7.3"
  },
  "scripts": {
    "build": "rollup -c",
    "dev": "yarn build --watch",
    "test": "jest --verbose",
    "test:watch": "yarn test --watch",
    "start-storybook": "start-storybook",
    "build-storybook": "build-storybook",
    "storybook": "start-storybook -p 6006",
    "chromatic": "npx chromatic --project-token=151685c386da"
  },
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {},
  "readme": "ERROR: No README data found!",
  "_id": "@lust/react@1.0.0"
}
