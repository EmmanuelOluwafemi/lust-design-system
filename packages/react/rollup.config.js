import Ts from 'rollup-plugin-typescript2'

export default {
    input: [
        'src/index.ts',
        'src/atoms/color/index.ts',
        'src/atoms/text/index.ts',
        'src/atoms/margin/index.ts',
        'src/molecules/select/index.ts'
    ],
    output: {
        dir: 'lib',
        format: 'esm',
        sourcemap: true
    },
    plugins: [Ts()],
    preserveModules: true,
    external: ['react', '@lust/foundation']
}