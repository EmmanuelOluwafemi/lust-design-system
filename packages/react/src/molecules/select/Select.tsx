import React, { createRef, KeyboardEventHandler, useEffect, useRef, useState } from 'react';

import Text from '../../atoms/text';

const KEY_CODES = {
    ENTER: "Enter",
    SPACE: " ",
    DOWN_ARROW: "ArrowDown",
    UP_ARROW: "ArrowUp",
    ESC: "Escape"
}

export interface SelectOption {
    label: string
    value: string
}

export interface RenderOptionProps  {
    isSelected: boolean
    option: SelectOption
    getOptionRecommendedProps: (overrideProps?: Object) => Object
}

export interface SelectProps {
    onOptionSelected?: (option: SelectOption, optionIndex: number) => void
    options?: SelectOption[]
    label?: string
    renderOption?: (props: RenderOptionProps) => React.ReactNode
}

const getPreviousOptionIndex = (currentIndex: number|null, options: Array<SelectOption>) => {
    if (currentIndex === null) {
        return 0
    }

    if (currentIndex === 0) {
        return options.length - 1
    }

    return currentIndex - 1
}

const getNextOptionIndex = (currentIndex: number|null, options: Array<SelectOption>) => {
    if (currentIndex === null) {
        return 0
    }

    if (currentIndex === options.length - 1) {
        return 0
    }

    return currentIndex + 1
}

const Select: React.FunctionComponent<SelectProps> = ({ options = [], label = 'Please select an option ...', onOptionSelected: handler, renderOption }) => {
    const [isOpen, setIsOpen] = useState<boolean>(false)
    const [selectedIndex, setSelectedIndex] = useState<null|number>(null)
    const [highlightedIndex, setHighlightedIndex] = useState<null|number>(null)
    const labelRef = useRef<HTMLButtonElement>(null);
    const [optionRefs, setOptionRefs] = useState<React.RefObject<HTMLLIElement>[]>([]);
    const [overlayTop, setOverlayTop] = useState<number>(0)

    const onOptionSelected = (option: SelectOption, optionIndex: number) => {
        if (handler) {
            handler(option, optionIndex)
        }

        setSelectedIndex(optionIndex)
        setIsOpen(false)
    }

    const onLabelClick = () => {
        setIsOpen(!isOpen)
    }

    useEffect(() => {
        setOverlayTop((
            labelRef.current?.offsetHeight || 0
        ) + 10)
    }, [labelRef.current?.offsetHeight])

    let selectedOption = null

    if (selectedIndex !== null) {
        selectedOption = options[selectedIndex]
    }

    const highlightOption = (optionIndex: number|null) => {
        setHighlightedIndex(optionIndex)
    }

    const onButtonKeyDown: KeyboardEventHandler = (event) => {
        event.preventDefault()

        if ([KEY_CODES.ENTER, KEY_CODES.SPACE, KEY_CODES.DOWN_ARROW].includes(event.key)) {
            setIsOpen(true)

            // set focus on the list item
            highlightOption(0)
        }
    }

    useEffect(() => {
        setOptionRefs(options.map(_ => createRef<HTMLLIElement>()))
    }, [options.length])

    useEffect(() => {
        if (highlightedIndex !== null && isOpen) {
            const ref = optionRefs[highlightedIndex]

            if (ref && ref.current) {
                ref.current.focus()
            }
        }
    }, [isOpen, highlightedIndex])

    const onOptionKeyDown: KeyboardEventHandler = (event) => {

        if (event.key === KEY_CODES.ESC) {
            setIsOpen(false)
            return
        }

        if (event.key === KEY_CODES.DOWN_ARROW) {
            highlightOption(getNextOptionIndex(highlightedIndex, options))
        }

        if (event.key === KEY_CODES.UP_ARROW) {
            highlightOption(
                getPreviousOptionIndex(highlightedIndex, options)
            )
        }

        if (event.key === KEY_CODES.ENTER) {
            onOptionSelected(options[highlightedIndex!], highlightedIndex!)
        }
    }

    return <div className='lust-select'>
        <button data-testid='LustSelectButton' onKeyDown={onButtonKeyDown} aria-controls='lust-select-list' aria-haspopup={true} aria-expanded={isOpen ? true: undefined} ref={labelRef} className='lust-select__label' onClick={() => onLabelClick()}>
            <Text>{selectedOption === null ? label : selectedOption.label}</Text>

            <svg className={`lust-select__caret ${isOpen ? 'lust-select__caret--open' : 'lust-select__caret--closed'}`} width='1rem' height='1rem' fill="none" strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} viewBox="0 0 24 24" stroke="currentColor"><path d="M19 9l-7 7-7-7" /></svg>
        </button>

        {(
            <ul role='menu' aria-hidden={isOpen ? undefined : false} id='lust-select-list' style={{ top: overlayTop }} className={`lust-select__dropdown ${isOpen ? 'lust-select__dropdown--open' : ''}`}>
                {options.map((option, optionIndex) => {
                    const isSelected = selectedIndex === optionIndex
                    const isHighlighted = highlightedIndex === optionIndex

                    const ref = optionRefs[optionIndex]

                    const renderOptionProps = {
                        ref,
                        option,
                        isSelected,
                        getOptionRecommendedProps: (overrideProps = {}) => {return {
                            ref,
                            role: 'menuitemradio',
                            'aria-label': option.label,
                            'aria-checked': isSelected ? true : undefined,
                            onKeyDown: onOptionKeyDown,
                            tabIndex: isHighlighted ? -1 : 0,
                            onMouseEnter: () => highlightOption(optionIndex),
                            onMouseLeave: () => highlightOption(null),
                            className: `lust-select__option
                                ${isSelected ? 'lust-select__option--selected' : ''}
                                ${isHighlighted ? 'lust-select__option--highlighted' : ''}
                            `,
                            key: option.value,
                            onClick: () => onOptionSelected(option, optionIndex),
                            ...overrideProps
                        }}
                    }

                    if (renderOption) {
                        return renderOption(renderOptionProps)
                    }

                    return <li
                        {...renderOptionProps.getOptionRecommendedProps()}
                        >
                            <Text>
                                {option.label}
                            </Text>

                            {isSelected ? (
                                <svg width='1rem' height='1rem' fill="none" strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} viewBox="0 0 24 24" stroke="currentColor"><path d="M5 13l4 4L19 7" /></svg>
                            ) : null}
                        </li>
                })}
            </ul>
        )}
    </div>
}

export default Select; 