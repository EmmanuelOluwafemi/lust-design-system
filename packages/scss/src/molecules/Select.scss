@import 'foundation/all';

.lust-select {
    display: flex;
    flex-direction: column;

    position: relative;

    &__label {
        display: flex;
        align-items: center;
        justify-content: space-between;

        @include padding('xs');
        background: $form-bg-color;
        border: 1px solid $form-border-color;
        
        font-family: $body-font-family ;

        cursor: pointer;
    }

    &__caret {
        transition:  all 250ms ease;

        &--open {
            transform: rotate(180deg);
        }

        &--close {
            transform: rotate(0deg);
        }
    }

    &__dropdown {
        @include padding('none');
        @include margin('none');

        border: 1px solid $form-border-color;
        background: $form-bg-color;
        z-index: 99;
        box-sizing: border-box;
        display: none;

        list-style-type: none;

        box-shadow: map-get($shadows, 'base');

        position: absolute;
        top: calc(100% + 5px);
        width: 100%;
    }

    &__dropdown--open {
        display: block;
    }

    &__option {
        display: flex;
        justify-content: space-between;
        align-items: center;
        cursor: pointer;
        @include padding('xs');
        box-sizing: border-box;
    }

    &__option--highlighted {
        background-color: $form-bg-color-hover;
    }

    &__option--selected {
        background-color: $form-bg-option-selected;
        color: $form-color-option-selected;
    }
}