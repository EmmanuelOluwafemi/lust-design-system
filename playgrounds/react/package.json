{
  "name": "@playgrounds/react",
  "version": "1.0.0",
  "main": "lib/index.js",
  "license": "MIT",
  "devDependencies": {
    "@types/react": "^18.0.12",
    "@types/react-dom": "^18.0.5",
    "parcel": "^2.6.0",
    "react": "^18.1.0",
    "react-dom": "^18.1.0",
    "typescript": "^4.7.3"
  },
  "scripts": {
    "dev": "parcel src/index.html -p 3000"
  },
  "private": "true"
}
