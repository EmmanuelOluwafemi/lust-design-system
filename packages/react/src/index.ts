import Color from './atoms/color'
import Text from './atoms/text'
import Margin from './atoms/margin'
import Select from './molecules/select'

export {
    Color,
    Text,
    Margin,
    Select
}