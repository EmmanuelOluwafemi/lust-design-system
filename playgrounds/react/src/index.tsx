import React from "react";
import ReactDOM from "react-dom/client";

import { Select } from "@lust/react"

import '@lust/scss/lib/Utilities.css'
import '@lust/scss/lib/Text.css'
import '@lust/scss/lib/Margin.css'
import '@lust/scss/lib/Select.css'
import '@lust/scss/lib/global.css'

const options = [
  { value: "1", label: "Option 1" },
  { value: "2", label: "Option 2" },
  { value: "3", label: "Option 3" },
  { value: "4", label: "Option 4" },
]

const root = ReactDOM.createRoot(
  document.getElementById("root") as HTMLElement,
);
root.render(
  <div style={{ padding: '40px' }}>
    <Select options={options} />
  </div>
);