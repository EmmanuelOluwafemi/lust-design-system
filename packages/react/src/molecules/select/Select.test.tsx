import React from "react";
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import Select from "./Select";

const options = [
  { value: "1", label: "Option 1" },
  { value: "2", label: "Option 2" },
  { value: "3", label: "Option 3" },
  { value: "4", label: "Option 4" },
]

test('renders all options passed to it', () => {
    const { getAllByRole, getByTestId } = render(<Select options={options} />)

    fireEvent.click(getByTestId('LustSelectButton'))

    expect(getAllByRole('menuitemradio')).toHaveLength(options.length)
})

test('renders options using custom renderOption method if passed as prop', () => {
  const { getAllByTestId } = render(<Select options={options} renderOption={({ option, getOptionRecommendedProps }) => {
    return <li data-testid='customRenderOption' {...getOptionRecommendedProps()}>{option.label}</li>
  }} />)

  expect(getAllByTestId('customRenderOption')).toHaveLength(options.length)
})

test('calls the onOptionSelected prop with the selected option and its index if passed', () => {
  const onOptionSelected = jest.fn()
  const { getAllByRole, getByTestId } = render(<Select options={options} onOptionSelected={onOptionSelected} />)

  fireEvent.click(getByTestId('LustSelectButton'))

  fireEvent.click(getAllByRole('menuitemradio')[0])

  expect(onOptionSelected).toHaveBeenCalledWith(options[0], 0)
})

test('the button label changes to the selected option label', () => {
  const { getAllByRole, getByTestId } = render(<Select options={options} />)

  fireEvent.click(getByTestId('LustSelectButton'))

  fireEvent.click(getAllByRole('menuitemradio')[0])

  expect(getByTestId('LustSelectButton')).toHaveTextContent(options[0].label)
})

test('snapshot of the selected option state', () => {
  const { getAllByRole, getByTestId, asFragment } = render(<Select options={options} />)

  fireEvent.click(getByTestId('LustSelectButton'))

  fireEvent.click(getAllByRole('menuitemradio')[0])

  expect(asFragment()).toMatchSnapshot()
})

test('snapshot of the base state', () => {
  const { asFragment } = render(<Select options={options} />)

  expect(asFragment()).toMatchSnapshot()
})

test('snapshot of the options menu open state', () => {
  const { getByTestId, asFragment } = render(<Select options={options} />)

  fireEvent.click(getByTestId('LustSelectButton'))

  expect(asFragment()).toMatchSnapshot()
})

test('can customize select label', () => {
  const { getByText } = render(<Select options={options} label="THIS IS A CUSTOM LABEL" />)

  expect(getByText(/THIS IS A CUSTOM LABEL/)).toBeInTheDocument()
})