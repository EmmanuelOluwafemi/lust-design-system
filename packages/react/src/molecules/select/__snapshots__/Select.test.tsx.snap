// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot of the base state 1`] = `
<DocumentFragment>
  <div
    class="lust-select"
  >
    <button
      aria-controls="lust-select-list"
      aria-haspopup="true"
      class="lust-select__label"
      data-testid="LustSelectButton"
    >
      <p
        class="kust-text lust-text-base"
      >
        Please select an option ...
      </p>
      <svg
        class="lust-select__caret lust-select__caret--closed"
        fill="none"
        height="1rem"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1rem"
      >
        <path
          d="M19 9l-7 7-7-7"
        />
      </svg>
    </button>
    <ul
      aria-hidden="false"
      class="lust-select__dropdown "
      id="lust-select-list"
      role="menu"
      style="top: 10px;"
    >
      <li
        aria-label="Option 1"
        class="lust-select__option
                                
                                
                            "
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="kust-text lust-text-base"
        >
          Option 1
        </p>
      </li>
      <li
        aria-label="Option 2"
        class="lust-select__option
                                
                                
                            "
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="kust-text lust-text-base"
        >
          Option 2
        </p>
      </li>
      <li
        aria-label="Option 3"
        class="lust-select__option
                                
                                
                            "
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="kust-text lust-text-base"
        >
          Option 3
        </p>
      </li>
      <li
        aria-label="Option 4"
        class="lust-select__option
                                
                                
                            "
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="kust-text lust-text-base"
        >
          Option 4
        </p>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`snapshot of the options menu open state 1`] = `
<DocumentFragment>
  <div
    class="lust-select"
  >
    <button
      aria-controls="lust-select-list"
      aria-expanded="true"
      aria-haspopup="true"
      class="lust-select__label"
      data-testid="LustSelectButton"
    >
      <p
        class="kust-text lust-text-base"
      >
        Please select an option ...
      </p>
      <svg
        class="lust-select__caret lust-select__caret--open"
        fill="none"
        height="1rem"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1rem"
      >
        <path
          d="M19 9l-7 7-7-7"
        />
      </svg>
    </button>
    <ul
      class="lust-select__dropdown lust-select__dropdown--open"
      id="lust-select-list"
      role="menu"
      style="top: 10px;"
    >
      <li
        aria-label="Option 1"
        class="lust-select__option
                                
                                
                            "
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="kust-text lust-text-base"
        >
          Option 1
        </p>
      </li>
      <li
        aria-label="Option 2"
        class="lust-select__option
                                
                                
                            "
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="kust-text lust-text-base"
        >
          Option 2
        </p>
      </li>
      <li
        aria-label="Option 3"
        class="lust-select__option
                                
                                
                            "
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="kust-text lust-text-base"
        >
          Option 3
        </p>
      </li>
      <li
        aria-label="Option 4"
        class="lust-select__option
                                
                                
                            "
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="kust-text lust-text-base"
        >
          Option 4
        </p>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`snapshot of the selected option state 1`] = `
<DocumentFragment>
  <div
    class="lust-select"
  >
    <button
      aria-controls="lust-select-list"
      aria-haspopup="true"
      class="lust-select__label"
      data-testid="LustSelectButton"
    >
      <p
        class="kust-text lust-text-base"
      >
        Option 1
      </p>
      <svg
        class="lust-select__caret lust-select__caret--closed"
        fill="none"
        height="1rem"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1rem"
      >
        <path
          d="M19 9l-7 7-7-7"
        />
      </svg>
    </button>
    <ul
      aria-hidden="false"
      class="lust-select__dropdown "
      id="lust-select-list"
      role="menu"
      style="top: 10px;"
    >
      <li
        aria-checked="true"
        aria-label="Option 1"
        class="lust-select__option
                                lust-select__option--selected
                                
                            "
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="kust-text lust-text-base"
        >
          Option 1
        </p>
        <svg
          fill="none"
          height="1rem"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1rem"
        >
          <path
            d="M5 13l4 4L19 7"
          />
        </svg>
      </li>
      <li
        aria-label="Option 2"
        class="lust-select__option
                                
                                
                            "
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="kust-text lust-text-base"
        >
          Option 2
        </p>
      </li>
      <li
        aria-label="Option 3"
        class="lust-select__option
                                
                                
                            "
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="kust-text lust-text-base"
        >
          Option 3
        </p>
      </li>
      <li
        aria-label="Option 4"
        class="lust-select__option
                                
                                
                            "
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="kust-text lust-text-base"
        >
          Option 4
        </p>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
